File: e-commerceAndroidApp/src/main/java/com/project/e_commerce/android/presentation/viewModel/SocialViewModel.kt
package com.project.e_commerce.android.presentation.viewModel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.project.e_commerce.domain.model.Post
import com.project.e_commerce.domain.model.Result
import com.project.e_commerce.domain.model.UserFollowModel
import com.project.e_commerce.domain.model.UserPost
import com.project.e_commerce.domain.model.UserProfile
import com.project.e_commerce.domain.usecase.post.CreatePostUseCase
import com.project.e_commerce.domain.usecase.user.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class SocialUiState(
    val isLoading: Boolean = false,
    val error: String? = null,
    val searchResults: List<UserProfile> = emptyList(),
    val followers: List<UserFollowModel> = emptyList(),
    val following: List<UserFollowModel> = emptyList(),
    val userPosts: List<UserPost> = emptyList(),
    val selectedUser: UserProfile? = null,
    val isFollowing: Boolean = false,
    val isFollowedBy: Boolean = false,
    val createdPost: Post? = null,
    val isCreatingPost: Boolean = false,
    val createPostError: String? = null
)

class SocialViewModel(
    private val createPostUseCase: CreatePostUseCase,
    private val followUserUseCase: FollowUserUseCase,
    private val unfollowUserUseCase: UnfollowUserUseCase,
    private val getFollowersUseCase: GetFollowersUseCase,
    private val getFollowingUseCase: GetFollowingUseCase,
    private val getFollowingStatusUseCase: GetFollowingStatusUseCase,
    private val getUserProfileUseCase: GetUserProfileUseCase,
    private val getUserPostsUseCase: GetUserPostsUseCase,
    private val updateUserProfileUseCase: UpdateUserProfileUseCase,
    private val searchUsersUseCase: SearchUsersUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(SocialUiState())
    val uiState: StateFlow<SocialUiState> = _uiState.asStateFlow()

    fun searchUsers(query: String) {
        if (query.isBlank()) {
            _uiState.update { it.copy(searchResults = emptyList()) }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            when (val result = searchUsersUseCase(query)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, searchResults = result.data) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }

    fun loadUserProfile(userId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            when (val result = getUserProfileUseCase(userId)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, selectedUser = result.data) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }

    fun loadUserPosts(userId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            when (val result = getUserPostsUseCase(userId)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, userPosts = result.data) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }

    fun checkFollowingStatus(currentUserId: String, targetUserId: String) {
        viewModelScope.launch {
            when (val result = getFollowingStatusUseCase(currentUserId, targetUserId)) {
                is Result.Success -> {
                    _uiState.update { 
                        it.copy(
                            isFollowing = result.data.isFollowing,
                            isFollowedBy = result.data.isFollowedBy
                        ) 
                    }
                }
                is Result.Error -> {
                    // Handle error silently or log
                }
                is Result.Loading -> { /* Loading state not needed here */ }
            }
        }
    }

    fun followUser(followerId: String, followedId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            when (val result = followUserUseCase(followerId, followedId)) {
                is Result.Success -> {
                     _uiState.update { it.copy(isLoading = false, isFollowing = true) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }

    fun unfollowUser(followerId: String, followedId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
             when (val result = unfollowUserUseCase(followerId, followedId)) {
                is Result.Success -> {
                     _uiState.update { it.copy(isLoading = false, isFollowing = false) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }

    fun loadFollowers(userId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            when (val result = getFollowersUseCase(userId)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, followers = result.data) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }

    fun loadFollowing(userId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            when (val result = getFollowingUseCase(userId)) {
                 is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false, following = result.data) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }

    fun updateUserProfile(userId: String, username: String, bio: String, photoUrl: String?) {
        viewModelScope.launch {
             _uiState.update { it.copy(isLoading = true) }
            val updatedProfile = UserProfile(
                uid = userId,
                username = username,
                bio = bio,
                profileImageUrl = photoUrl
            )
            when (val result = updateUserProfileUseCase(updatedProfile)) {
                is Result.Success -> {
                    _uiState.update { it.copy(isLoading = false) }
                }
                is Result.Error -> {
                    _uiState.update { it.copy(isLoading = false, error = result.error.message) }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }
    
    fun selectUser(user: UserProfile) {
        _uiState.update { it.copy(selectedUser = user) }
    }
    
    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
    
    /**
     * Create a new post/reel
     * @param type "reel", "product", or "photo"
     * @param mediaUrl URL of uploaded media
     * @param caption Optional description
     */
    fun createPost(type: String, mediaUrl: String, caption: String?) {
        viewModelScope.launch {
            _uiState.update { it.copy(isCreatingPost = true, createPostError = null, createdPost = null) }
            when (val result = createPostUseCase(type, mediaUrl, caption)) {
                is Result.Success -> {
                    _uiState.update { 
                        it.copy(
                            isCreatingPost = false, 
                            createdPost = result.data,
                            createPostError = null
                        ) 
                    }
                }
                is Result.Error -> {
                    _uiState.update { 
                        it.copy(
                            isCreatingPost = false, 
                            createPostError = result.error.message
                        ) 
                    }
                }
                is Result.Loading -> { /* Already showing loading state */ }
            }
        }
    }
    
    /**
     * Clear created post state (after navigation or display)
     */
    fun clearCreatedPost() {
        _uiState.update { it.copy(createdPost = null, createPostError = null) }
    }
}

File: e-commerceiosApp/e-commerceiosApp/ViewModels/SocialViewModel.swift
import Foundation
import Shared

class SocialViewModel: ObservableObject {
    @Published var searchResults: [UserProfile] = []
    @Published var followers: [UserProfile] = []
    @Published var following: [UserProfile] = []
    
    // Pour le profil visité
    @Published var userPosts: [UserPost] = []
    @Published var bookmarkedPosts: [UserPost] = []
    @Published var likedPosts: [UserPost] = []
    @Published var isFollowing: Bool = false
    @Published var selectedUser: UserProfile?
    
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    // Computed
    var totalLikes: Int {
        userPosts.reduce(0) { $0 + Int($1.likesCount) }
    }
    
    // Dependencies
    private let followUserUseCase = DependencyWrapper.shared.followUserUseCase
    private let unfollowUserUseCase = DependencyWrapper.shared.unfollowUserUseCase
    private let getFollowersUseCase = DependencyWrapper.shared.getFollowersUseCase
    private let getFollowingUseCase = DependencyWrapper.shared.getFollowingUseCase
    private let getFollowingStatusUseCase = DependencyWrapper.shared.getFollowingStatusUseCase
    private let getUserPostsUseCase = DependencyWrapper.shared.getUserPostsUseCase
    private let updateUserProfileUseCase = DependencyWrapper.shared.updateUserProfileUseCase
    private let searchUsersUseCase = DependencyWrapper.shared.searchUsersUseCase
    private let getLikedPostsUseCase = DependencyWrapper.shared.getLikedPostsUseCase
    private let getBookmarkedPostsUseCase = DependencyWrapper.shared.getBookmarkedPostsUseCase
    
    // Recherche
    func searchUsers(query: String) {
        guard !query.isEmpty else {
            searchResults = []
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        searchUsersUseCase.invoke(query: query) { result, error in
            DispatchQueue.main.async {
                self.isLoading = false
                if let success = result as? ResultSuccess<NSArray> {
                     if let users = success.data as? [UserProfile] {
                         self.searchResults = users
                     }
                } else if let errorResult = result as? ResultError {
                    // Ignorer erreur temporaire pour demo ou afficher
                    self.errorMessage = errorResult.error.message
                } else if let error = error {
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }
    
    // Actions Follow
    func followUser(targetId: String) {
        let currentUserId = SessionManager.shared.currentUserId
        isLoading = true
        
        followUserUseCase.invoke(followerId: currentUserId, followedId: targetId) { result, error in
            DispatchQueue.main.async {
                self.isLoading = false
                if result is ResultSuccess<KotlinUnit> {
                    self.isFollowing = true
                    // Update stats if needed locally
                } else if let errorResult = result as? ResultError {
                    self.errorMessage = errorResult.error.message
                }
            }
        }
    }
    
    func unfollowUser(targetId: String) {
        let currentUserId = SessionManager.shared.currentUserId
        isLoading = true
        
        unfollowUserUseCase.invoke(followerId: currentUserId, followedId: targetId) { result, error in
            DispatchQueue.main.async {
                self.isLoading = false
                 if result is ResultSuccess<KotlinUnit> {
                    self.isFollowing = false
                } else if let errorResult = result as? ResultError {
                    self.errorMessage = errorResult.error.message
                }
            }
        }
    }
    
    func checkFollowingStatus(targetId: String) {
        let currentUserId = SessionManager.shared.currentUserId
        
        getFollowingStatusUseCase.invoke(followerId: currentUserId, followedId: targetId) { result, error in
            DispatchQueue.main.async {
                if let success = result as? ResultSuccess<KotlinBoolean> {
                    self.isFollowing = success.data as? Bool ?? false
                }
            }
        }
    }
    
    // Data Loading
    func loadUserPosts(userId: String) {
        isLoading = true
        getUserPostsUseCase.invoke(userId: userId) { result, error in
            DispatchQueue.main.async {
                self.isLoading = false
                if let success = result as? ResultSuccess<NSArray> {
                    if let posts = success.data as? [UserPost] {
                        self.userPosts = posts
                    }
                } else if let errorResult = result as? ResultError {
                    // "Not implemented" likely
                    self.errorMessage = errorResult.error.message
                }
            }
        }
    }
    
    func loadFollowers(userId: String) {
        isLoading = true
        getFollowersUseCase.invoke(userId: userId) { result, error in
            DispatchQueue.main.async {
                self.isLoading = false
                 if let success = result as? ResultSuccess<NSArray> {
                    if let users = success.data as? [UserProfile] {
                        self.followers = users
                    }
                }
            }
        }
    }
    
    func loadFollowing(userId: String) {
        isLoading = true
        getFollowingUseCase.invoke(userId: userId) { result, error in
             DispatchQueue.main.async {
                self.isLoading = false
                 if let success = result as? ResultSuccess<NSArray> {
                    if let users = success.data as? [UserProfile] {
                        self.following = users
                    }
                }
            }
        }
    }
    
    func updateUserProfile(username: String, bio: String, photoUrl: String?, completion: @escaping (Bool) -> Void) {
        let userId = SessionManager.shared.currentUserId
        isLoading = true
        
        updateUserProfileUseCase.invoke(userId: userId, username: username, bio: bio, photoUrl: photoUrl) { result, error in
            DispatchQueue.main.async {
                self.isLoading = false
                if result is ResultSuccess<UserProfile> {
                    completion(true)
                } else {
                    if let errorResult = result as? ResultError {
                        self.errorMessage = errorResult.error.message
                    }
                    completion(false)
                }
            }
        }
    }
    
    // Utils
    func selectUser(_ user: UserProfile) {
        self.selectedUser = user
        self.userPosts = []
        self.checkFollowingStatus(targetId: user.uid)
        self.loadUserPosts(userId: user.uid)
    }
    
    // MARK: - Bookmarked Posts
    func loadBookmarkedPosts(userId: String) {
        getBookmarkedPostsUseCase.invoke(userId: userId) { result, error in
            DispatchQueue.main.async {
                if let success = result as? ResultSuccess<NSArray> {
                    if let posts = success.data as? [Post] {
                        self.bookmarkedPosts = posts.map { post in
                            UserPost(
                                id: post.id,
                                userId: post.userId,
                                type: "REEL",
                                title: post.caption,
                                description: post.caption,
                                mediaUrl: post.mediaUrl,
                                thumbnailUrl: post.thumbnailUrl,
                                images: [],
                                likesCount: Int32(post.likesCount),
                                commentsCount: Int32(post.commentsCount),
                                viewsCount: 0,
                                isPublished: true,
                                createdAt: post.createdAt,
                                updatedAt: post.updatedAt
                            )
                        }
                    }
                }
            }
        }
    }
    
    // MARK: - Liked Posts
    func loadLikedPosts(userId: String) {
        getLikedPostsUseCase.invoke(userId: userId) { result, error in
            DispatchQueue.main.async {
                if let success = result as? ResultSuccess<NSArray> {
                    if let posts = success.data as? [Post] {
                        self.likedPosts = posts.map { post in
                            UserPost(
                                id: post.id,
                                userId: post.userId,
                                type: "REEL",
                                title: post.caption,
                                description: post.caption,
                                mediaUrl: post.mediaUrl,
                                thumbnailUrl: post.thumbnailUrl,
                                images: [],
                                likesCount: Int32(post.likesCount),
                                commentsCount: Int32(post.commentsCount),
                                viewsCount: 0,
                                isPublished: true,
                                createdAt: post.createdAt,
                                updatedAt: post.updatedAt
                            )
                        }
                    }
                }
            }
        }
    }
}
